server/etcdserver/api/membership/cluster.go

```diff
@@ -303,9 +303,27 @@ func (c *RaftCluster) Recover(onSet func(*zap.Logger, *semver.Version)) {
 
 // ValidateConfigurationChange takes a proposed ConfChange and
 // ensures that it is still valid.
-func (c *RaftCluster) ValidateConfigurationChange(cc raftpb.ConfChange) error {
+func (c *RaftCluster) ValidateConfigurationChange(cc raftpb.ConfChange, shouldApplyV3 ShouldApplyV3) error {
 	// TODO: this must be switched to backend as well.
-	membersMap, removedMap := membersFromStore(c.lg, c.v2store)
+	var membersMap map[types.ID]*Member
+	var removedMap map[types.ID]bool
+	if c.v2store != nil {
```

Comment:
Yes, it's true that we only need to verify v3 just as I mentioned in https://github.com/etcd-io/etcd/pull/16084#issuecomment-1698814517. But just as I mentioned in one of previous community meetings, we should backport the change to 3.5, so let's keep the v2 validation for now. After the backport is done, we can remove it later. Please also see https://github.com/etcd-io/etcd/pull/16084#issuecomment-1696362494

Reply:
I'm not againt fixing 3.5 with backport in general, but it doesn't change our assumptions... 
Someone can rollback from 3.6.x to 3.5.0 and experience consequences of V2 store being stale (but still considered as source of truth). Seems we would need another safeguard that invalidates Store V2 in etcd 3.6 such that rollback to version that 'does not understand' the invalidation is not possible...

Reply:
>  experience consequences of V2 store being stale (but still considered as source of truth)

Should we change the source of truth to V3 store (bbolt) in 3.5? I prefer to YES.

Reply:
> but it doesn't change our assumptions...
> Someone can rollback from 3.6.x to 3.5.0 and experience consequences of V2 store being stale (but still considered as source of truth). 
> 

Yes, it's true that the v2 store of **3.5** might be stale right after bootstrapping (no matter it's downgraded from 3.6, or just 3.5 restarts), because 3.5 loads membership data from snapshot which might be stale; but I do not see any issue.
- Eventually the membership data in v2 store will be in sync with v3 store when the replaying WAL entries is done;
- Note etcd 3.5 isn't able to serve client requests until it [finishes publishing the server info to cluster through raft](https://github.com/etcd-io/etcd/blob/v3.5.9/server/embed/serve.go#L101), which must be applied after the replaying legacy WAL entries.

So 
- My previous comment https://github.com/etcd-io/etcd/pull/16084#issuecomment-1696362494 is also invalid.
- We don't have to change the [source of truth for membership data](https://github.com/etcd-io/etcd/blob/v3.5.9/server/etcdserver/api/membership/cluster.go#L259-L265) to V3 store (bbolt) in **3.5**.

@ptabor please kindly let me know if I misunderstood your point or you have any other concern.

Reply:
@ahrtr you are missing one important consequence of changing validation. Normal raft proposal (like PUT) can be represented as a state machine without side effect. It means that we can care about only the end state, so assuming that we all members execute same WAL entries, we are good. This however is not true for membership.

Membership "state machine" has side effect in form of ConfState, which is state of truth for raft voting members. When applying the entry we run the validation function against the membership info to decide if it should have an effect on ConfState. This means that we not only care about the state, but also transitions. Every time we replay WAL we need to guarantee that validate behaves the same. That also applies to +1/-1 etcd versions to support upgrades.

Reply:
In term of "side effect on ConfState", I assume you are talking about [server.go#L1989](https://github.com/etcd-io/etcd/blob/e85949db3a05dc7f746aa8c6c6416a9822d2a779/server/etcdserver/server.go#L1989) or [server.go#L2001](https://github.com/etcd-io/etcd/blob/e85949db3a05dc7f746aa8c6c6416a9822d2a779/server/etcdserver/server.go#L2001). Pls let me know if we are not on the same page. 

Note that I don't think this PR changes the validation logic on membership data. It just applies the validation on v3 store as well. If the validation on v2 store fails (or succeeds) [Of course, we will remove the validation on v2], then the validation on v3 should fails (or succeeds) as well. 

Pls be more specific if you still think there is any issue.

Again, just as I mentioned in https://github.com/etcd-io/etcd/pull/16084#issuecomment-1734036435, since this is the most important part, so any suggestion on how to improve the test cases is welcome.